epc_data_analysis/pipeline/feature_engineering.py

```diff
@@ -0,0 +1,279 @@
+    df["CURR_ENERGY_RATING_NUM"] = df.CURRENT_ENERGY_RATING.map(rating_dict)
+
+    # EPC rating in category (A-B, C-D or E-G)
+    df["ENERGY_RATING_CAT"] = df.CURRENT_ENERGY_RATING.map(EPC_cat_dict)
+
+    # Numerical difference between current and potential energy rating (A-G)
+    df["DIFF_POT_ENERGY_RATING"] = (
+        df.POTENTIAL_ENERGY_RATING.map(rating_dict) - df["CURR_ENERGY_RATING_NUM"]
+    )
+
+    # Set DIFF_POT_ENERGY_RATING to 0.0
+    # if substraction yielded value below 0.0 due to input error (few cases).
+    df = df[df.DIFF_POT_ENERGY_RATING >= 0.0]
+
+    return df
+
+
+def map_quality_to_number(df, list_of_features):
+
+    quality_to_num_dict = {
+        "Very Good": 5.0,
+        "Good": 4.0,
+        "Average": 3.0,
+        "Poor": 2.0,
+        "Very Poor": 1.0,
+    }
+
+    for feature in list_of_features:
+        df[feature + "_AS_NUM"] = df[feature].map(quality_to_num_dict)
+
+    return df
+
+
+def get_heating_features(df, fine_grained_HP_types=False):
+    """Get heating type category based on HEATING_TYPE category.
+    heating_system: heat pump, boiler, community scheme etc.
+    heating_source: oil, gas, LPC, electric.
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+        Dataframe that is updated with heating features.
+
+    fine_grained_HP_types : bool, default=False
+        If True, get different heat pump types (air sourced, ground sourced etc.).
+        If False, return "heat pump" as heating type category.
+
+    Return
+    ---------
+    df : pandas.DataFrame
+        Updated dataframe with heating system and source."""
+
+    general_type = []
+    heating_type = []
+
+    heating_types = df["MAINHEAT_DESCRIPTION"]
+
+    # Get specific and general heating category for each entry
+    for heating in heating_types:
+
+        if not pd.isnull(heating):
+
+            # Lowercase
+            heating = heating.lower()
+
+            # Different heat pump types
+
+            if "ground source heat pump" in heating:
+                h_type = "ground source heat pump"
+                g_type = "electric"
+
+            elif "air source heat pump" in heating:
+                h_type = "air source heat pump"
+                g_type = "electric"
+
+            elif "water source heat pump" in heating:
+                h_type = "water source heat pump"
+                g_type = "electric"
+
+            elif "heat pump" in heating:
+                h_type = "heat pump"
+                g_type = "electric"
+
+            # Electric heaters
+
+            elif "electric storage heaters" in heating:
+                h_type = "storage heater"
+                g_type = "electric"
+
+            elif "electric underfloor heating" in heating:
+                h_type = "underfloor heating"
+                g_type = "electric"
+
+            # Boiler and radiator
+
+            elif "boiler and radiator" in heating:
+                if "gas" in heating:
```

Comment:
I feel like this long list of if statements could be made easier to read and edit by the use of dictionaries.
e.g.
```
elif "boiler and radiator" in heating:
    h_type = "boiler and radiator"
    g_type = 'unknown'
    g_type_dict = {'gas': 'gas', ', oil': 'oil', 'lpg': 'LPG', 'electric': 'electric'}
    for word, g_type_name in g_type_dict.items():
        if word in heating:
            g_type = g_type_name
elif "boiler and underfloor" in heating or "boiler & underfloor" in heating:
 ....
```